### ISSUE TITLE
Add a running example of iframe testing with Cypress

### ISSUE THREAD

**Issue by dmundra** [CivicActions/cypress-tests#43]
Maybe submodule this ckeditor demo https://github.com/ckeditor/ckeditor5-demos/tree/master/feature-rich and then write tests that interact with it?

**Comment by alexfinnarn** [comment]
I've seen and used code to test ckeditor 4 like:

```js
Cypress.Commands.add('getCKEditorBody', () => {
  // get the iframe > document > body
  // and retry until the body element is not empty

  return cy
  .get('iframe.cke_wysiwyg_frame')
  .its('0.contentDocument.body').should('not.be.empty')
  // wraps "body" DOM element to allow
  // chaining more Cypress commands, like ".find(...)"
  // https://on.cypress.io/wrap
  .then((body) => cy.wrap(body))
});
```

but for some iframes, I haven't gotten any of the Cypress iframe example code to work since the iframe loads asynchronous content and Cypress captures an empty DOM.

But the bigger question here is how to create an app that allows us to write example tests against it. My thought was to use a simple framework with simple HTML and including ckeditor 5 could fit into a route with only a WYSIWYG or as part of a form upload process.

Since Cypress is JS-based, I think using a JS-based backend framework like Express would allow for the most use of Cypress and features like spying on function calls.

Have you thought about adding an app to this project? I would say it could be static HTML, but then testing authentication would be weird and probably want to reach for an actual backend at some point anyway.

**Comment by dmundra** [comment]
> I haven't gotten any of the Cypress iframe example code to work since the iframe loads asynchronous content and Cypress captures an empty DOM.

I wonder if we can use an `intercept` and a `wait` to make that work?

> Have you thought about adding an app to this project? I would say it could be static HTML, but then testing authentication would be weird and probably want to reach for an actual backend at some point anyway.

The cypress demo example is using webpack. We might be able to use webpack for our own examples as well and see how that works in the pipeline? I was going to give that a try for this issue.

Another option is a jekyll based site like https://accessibility.civicactions.com/. I think we can test most everything with static sites except authentication unless we want to setup a simple JS auth app with fake data. I haven't give that much thought.

**Comment by alexfinnarn** [comment]
> I wonder if we can use an intercept and a wait to make that work?

The iframe I'm referencing is on https://www.cms.gov/newsroom at the bottom with the three Twitter widgets. I was simply trying to assert something loads inside, like the Twitter handle title. I tried several solutions in https://github.com/cypress-io/cypress/issues/136 but nothing consistently worked. Your suggestion might work, but I moved on from trying to write a test for those Twitter widgets.

As far as creating an app to test on, I still think Express might be the best option, and I don't think we'd need Webpack if we're not bundling modules to be used in the frontend...I don't follow it closely, but apparently you can now import node dependencies natively in JS files with modern browsers so hopefully we can keep the dependencies low and the testing app very simple.

I will create an issue to explore using Express and see if I can set up a page that loads ckeditor. Then this issue could use that code to write an example test against. Sound like a plan?

**Comment by dmundra** [comment]
That works for me! Thank you.

**Comment by alexfinnarn** [comment]
I created https://github.com/CivicActions/cypress-tests/pull/44 which has a "/ckeditor" page to run tests against. It uses the "kitchen sink" version of CKEditor, because I thought that would provide the most robust version to test against...but a simpler build might make more sense. I also chose Twig for the templating since I think CAers will be more familiar with that templating engine, but handlebars is also an option. Overall, it seems like Express is a good candidate to keep things simple but provide all the functionality needed to create example tests.
